File: JavaLibrary/src/test/java/com/flat/simplex/support/LineCallChain.java
package com.flat.simplex.support;

import com.flat.simplex.lexer.Key;
import com.flat.simplex.lexer.Token;
import com.flat.simplex.parser.logic.line.*;
import com.flat.simplex.parser.logic.line.call.LineCall;
import org.opentest4j.AssertionFailedError;

import java.util.ArrayList;

import static org.junit.jupiter.api.Assertions.assertEquals;

public class LineCallChain {

    ArrayList<Object> objects = new ArrayList<>();

    public static LineCallChain lChain(Class... classes) {
        return new LineCallChain().add(classes);
    }

    public static LineCallChain lChain(Key op) {
        return new LineCallChain().add(op);
    }

    public static LineCallChain lChain(LineCallChain cChain) {
        return new LineCallChain().add(cChain);
    }

    public LineCallChain add(Class... classes) {
        objects.add(classes);
        return this;
    }

    public LineCallChain add(Key op) {
        objects.add(op);
        return this;
    }

    public LineCallChain add(LineCallChain cChain) {
        objects.add(cChain);
        return this;
    }

    public void assertChain(LineValue lineValue, String message) {
        if (lineValue instanceof LineChain) {
            if (objects.size() != 1) {
                throw new AssertionFailedError(message, "LineGroup", "LineChain");
            } else {
                Class[] classes = (Class[]) objects.get(0);
                LineChain lc = (LineChain) lineValue;
                assertLocalLineChain(message, classes, lc);
            }
        } else {
            LineGroup lc = (LineGroup) lineValue;
            assertEquals(objects.size(), lc.getLines().size(), "Invalid chain size");
            for (int i = 0; i < objects.size(); i++) {
                Line line = lc.getLines().get(i);
                Object obj = objects.get(i);
                if (obj instanceof Class[]) {
                    if (line instanceof LineChain) {
                        assertLocalLineChain(message, (Class[])obj, (LineChain) line);
                    } else {
                        throw new AssertionFailedError(message, "LineChain", line.getClass().getSimpleName());
                    }
                } else if (obj instanceof LineCallChain) {
                    if (line instanceof LineGroup) {
                        ((LineCallChain) obj).assertChain((LineGroup)line, message);
                    } else {
                        throw new AssertionFailedError(message, "LineGroup", line.getClass().getSimpleName());
                    }
                } else if (obj instanceof Key) {
                    if (line instanceof LineOp) {
                        assertEquals((Key)obj, ((LineOp) line).getKey(), message);
                    } else {
                        throw new AssertionFailedError(message, "LineOp", line.getClass().getSimpleName());
                    }
                }
            }
        }
    }

    private void assertLocalLineChain(String message, Class[] classes, LineChain lc) {
        LineCall call = lc.getFirstCall();
        int i = 0;
        while (call != null) {
            if (i >= classes.length) {
                throw new AssertionFailedError(message, classes.length  + " members", (i + 1) + " members");
            }
            if (call.getClass() != classes[i]) {
                throw new AssertionFailedError(message, classes[i].getSimpleName(), call.getClass().getSimpleName());
            }
            i++;
            call = call.getNext();
        }
        if (i < classes.length) {
            throw new AssertionFailedError(message, classes.length  + " members", i + " members");
        }

    }

    private static class ClassChain {
        public Class<?> clazz;
        public ClassChain next;
        public ClassChain child;

        public ClassChain(Class<?> clazz) {
            this.clazz = clazz;
        }
    }
}

File: SimplexApp/simplex/support/LineCallChain.js
const AssertError = require("simplex/support/AssertError.js");
const Key = require("simplex/lexer/Key.js");
const LineChain = require("simplex/parser/logic/line/LineChain.js");
const LineGroup = require("simplex/parser/logic/line/LineGroup.js");
const LineOp = require("simplex/parser/logic/line/LineOp.js");

class LineCallChain {
    constructor() {
        this.objects = [];
    }
    
    add(cChain, ...classes) {
        if (typeof cChain === 'string') {
            if (classes && classes.length > 0) {
                this.objects.push([cChain].concat(classes));
            } else {
                this.objects.push([cChain]);
            }
        } else {
            this.objects.push(cChain);
        }
        return this;
    }
    
    assertChain(lineValue, message) {
        if (lineValue instanceof LineChain) {
            if (this.objects.length !== 1) {
                throw new AssertError(message, 'LineGroup', 'LineChain');
            } else {
                let classes = this.objects[0];
                let lc = lineValue;
                this.assertLocalLineChain(message, classes, lc);
            }
        } else if (lineValue instanceof LineGroup) {
            let lc = lineValue;
            expect(this.objects.length, 'Invalid chain size').toBe(lc.getLines().length);

            for (let i = 0; i < this.objects.length; i++) {
                let line = lc.getLines()[i];
                let obj = this.objects[i];
                if (Array.isArray(obj)) {
                    if (line instanceof LineChain) {
                        this.assertLocalLineChain(message, obj, line);
                    } else {
                        throw new AssertError(message, 'LineChain', line.constructor.name);
                    }
                } else if (obj instanceof LineCallChain) {
                    if (line instanceof LineGroup) {
                        (obj).assertChain(line, message);
                    } else {
                        throw new AssertError(message, 'LineGroup', line.constructor.name);
                    }
                } else if (obj instanceof Key) {
                    if (line instanceof LineOp) {
                        expect(obj, message).toBe(line.getKey());
                    } else {
                        throw new AssertError(message, 'LineOp', line.constructor.name);
                    }
                }
            }
        } else if (this.objects.length === 0) {
            expect(lineValue, message).toBeNull();
        } else {
            throw new AssertError(message, '(Not Empty)', '(Empty)');
        }
    }

    assertLocalLineChain(message, classes, lc) {
        let call = lc.getFirstCall();
        let i = 0;
        while (call !== null) {
            if (i >= classes.length) {
                throw new AssertError(message, classes.length + ' members', (i + 1) + ' members');
            }
            if (call.constructor.name !== classes[i]) {
                throw new AssertError(message, classes[i], call.constructor.name);
            }
            i++;
            call = call.getNext();
        }
        if (i < classes.length) {
            throw new AssertError(message, classes.length + ' members', i + ' members');
        }
    }

    static lChain (cChain, ...classes) {
        return new LineCallChain().add(cChain, ...classes);
    }
}

module.exports = LineCallChain;